finagle/buoyant/src/main/scala/com/twitter/finagle/buoyant/ExistentialStability.scala

```diff
@@ -56,44 +65,63 @@ object ExistentialStability {
   implicit class ExistentialAct[T](val unstable: Activity[Option[T]])
     extends AnyVal {
     def stabilizeExistence: Activity[Option[Var[T]]] = {
-      val inner = Var.async[Activity.State[Option[VarUp[T]]]](Activity.Pending) { update =>
-        // the current inner Var, null if the outer Var is None
-        @volatile var current: VarUp[T] = null
-        @volatile var exists = false
-        val mu = new {}
+
+      // The inner Var is null until the first time unstable becomes a Some.
+      @volatile var inner: Var[T] = null
+
+      def makeInner(init: T) = Var.async(init) { update =>
         unstable.states.respond {
-          case Activity.Ok(Some(t)) if current == null => mu.synchronized {
-            // T created
-            current = Var(t)
-            exists = true
-            update() = Activity.Ok(Some(current))
-          }
-          case Activity.Ok(Some(t)) if !exists => mu.synchronized {
-            // T recreated
-            exists = true
-            current() = t
-            update() = Activity.Ok(Some(current))
-          }
-          case Activity.Ok(Some(t)) => mu.synchronized {
-            // T modified
-            current() = t
-          }
-          case Activity.Ok(None) => mu.synchronized {
-            // T deleted
-            exists = false
-            update() = Activity.Ok(None)
-          }
-          case Activity.Pending => mu.synchronized {
-            update() = Activity.Pending
-            exists = false
-          }
-          case Activity.Failed(e) => mu.synchronized {
-            update() = Activity.Failed(e)
-            exists = false
+          case Activity.Ok(Some(tt)) => update() = tt
+          case _ =>
+        }
+      }
+
+      // This keeps track of the state of the outer Var:
+      // None means the state was not Activity.Ok
+      // Some(false) means the state was Activity.Ok(Some)
+      // Some(true) means the state was Activity.Ok(None)
+      //
+      // We need to track these three states because we only want to update the outer Var when we
+      // transition between states.
+      var exists: Option[Boolean] = None
+      val mu = new {}
+
+      val outer = Var.async[Activity.State[Option[Var[T]]]](Activity.Pending) { update =>
+        unstable.states.respond { state =>
+          mu.synchronized {
+            state match {
+              case Activity.Ok(Some(t)) if inner == null =>
+                // T created.
+                inner = makeInner(t)
+                exists = Some(true)
+                update() = Activity.Ok(Some(inner))
+              case Activity.Ok(Some(_)) if exists != Some(true) =>
+                // T re-created.
+                exists = Some(true)
+                update() = Activity.Ok(Some(inner))
+              case Activity.Ok(Some(_)) if exists == Some(true) =>
```

Comment:
tioli: `exists.getOrElse(false)` and `!exists.getOrElse(true)` are a bit easier to read than e.g. `exists != Some(false)`

Reply:
ah but that's not the same thing, because the None type.  Consider using an dedicated marker `enum` with three stats rather than `Option[Boolean]`

Reply:
I considered this, but though it might be more boilerplate than necessary.  But you're probably right, it will be more clear.

Reply:
It may be, but feel free to disregard if it's not

Reply:
I opted for a sealed trait because scala enums give me the heebie jeebies